opshin/type_inference.py

```diff
@@ -1010,6 +1010,29 @@ def visit_FunctionDef(self, node: FunctionDef) -> TypedFunctionDef:
     def visit_Module(self, node: Module) -> TypedModule:
         self.enter_scope()
         tm = copy(node)
+        
+        # First pass: extract all function signatures and add them to scope
+        # This enables mutual recursion by making all functions available before processing bodies
+        for stmt in node.body:
+            if isinstance(stmt, ast.FunctionDef):
+                try:
+                    # Create a minimal function type signature from the annotation
+                    functyp = FunctionType(
```

Comment:
this currently works if no user defined classes are in play. however, it could happen that user defined classes are added in between. this needs to be taken into account (add a test case for this!)